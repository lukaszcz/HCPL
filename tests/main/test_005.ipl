
// call by name test

let g = \#x { print x; g };

let f0 = \#x \!y {
   if = x 0 then
      y
   else {
      print 1;
      f0 (- x 1) (y 0)
   }
};

let f1 = \#x \#y {
   if = x 0 then
      y
   else {
      print 1;
      f1 (- x 1) (y 0)
   }
};

let h = \#x { print 7; print x; + x x; };
let he = \!x { print 7; print x; + x x; };

f0 5 g;
print 7;
f1 5 g;
print (h {print 2; 3});
print (he {print 2; 3});
print (he #{print 2; 3});

let! x = {print 8; 9};
print (h x);

let# y = {print 8; 9};
print (h y);

print (h {print 8; 9});
