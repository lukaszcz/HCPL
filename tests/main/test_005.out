1
0
1
0
1
0
1
0
1
0
7
1
1
1
1
1
0
0
0
0
0
7
2
3
2
2
6
2
7
3
6
7
2
3
2
2
6
8
7
9
18
7
8
9
8
8
18
7
8
9
8
8
18
